More CLAAS-3 improvements

sfinkens opened the pull request:
Based on feedback from Jan-Fokke.

Here is an example ncdump:

<details>

```
netcdf S_NWC_seviri_meteosat9_99999_20090701T1200000Z_20090701T1215000Z {
dimensions:
        time = UNLIMITED ; // (1 currently)
        x = 3712 ;
        y = 3712 ;
        corners = 4 ;
        bnds_1d = 2 ;
variables:
        double time(time) ;
                time:bounds = "time_bnds" ;
                time:standard_name = "time" ;
                time:units = "days since 2004-01-01" ;
                time:calendar = "standard" ;
        double x(x) ;
                x:standard_name = "projection_x_coordinate" ;
                x:units = "m" ;
        double y(y) ;
                y:standard_name = "projection_y_coordinate" ;
                y:units = "m" ;
        float lon(y, x) ;
                lon:_FillValue = -999.f ;
                lon:end_time = "2009-07-01 12:12:40.964000" ;
                lon:long_name = "longitude coordinate" ;
                lon:modifiers = "" ;
                lon:standard_name = "longitude" ;
                lon:start_time = "2009-07-01 12:00:10.021000" ;
                lon:units = "degrees_east" ;
        float lat(y, x) ;
                lat:_FillValue = -999.f ;
                lat:end_time = "2009-07-01 12:12:40.964000" ;
                lat:long_name = "latitude coordinate" ;
                lat:modifiers = "" ;
                lat:standard_name = "latitude" ;
                lat:start_time = "2009-07-01 12:00:10.021000" ;
                lat:units = "degrees_north" ;
        double acq_time(y) ;
                acq_time:_FillValue = -9999. ;
                acq_time:units = "milliseconds since 2009-07-01T12:00:00" ;
                acq_time:calendar = "standard" ;
        short azimuthdiff(time, y, x) ;
                azimuthdiff:_FillValue = -32767s ;
                azimuthdiff:end_time = "2009-07-01 12:12:40.964000" ;
                azimuthdiff:id_tag = "azimuthdiff" ;
                azimuthdiff:long_name = "absolute azimuth difference angle" ;
                azimuthdiff:modifiers = "" ;
                azimuthdiff:standard_name = "absolute_angle_of_rotation_from_solar_azimuth_to_platform_azimuth" ;
                azimuthdiff:start_time = "2009-07-01 12:00:10.021000" ;
                azimuthdiff:units = "degree" ;
                azimuthdiff:valid_range = 0LL, 18000LL ;
                azimuthdiff:coordinates = "acq_time lat lon" ;
                azimuthdiff:add_offset = 0. ;
                azimuthdiff:scale_factor = 0.01 ;
        byte georef_offset_corrected(time) ;
                georef_offset_corrected:comment = "Until December 2017, SEVIRI L1.5 data is shifted by 1.5km SSP North and West against the nominal GEOS projection. Since December 2017 this offset has been corrected by EUMETSAT. This flag indicates if the correction has been applied. If not, the area extent is adjusted so that the projection always matches the data." ;
                georef_offset_corrected:flag_meanings = "georef_offset_present georef_offset_corrected" ;
                georef_offset_corrected:flag_values = 0b, 1b ;
                georef_offset_corrected:long_name = "Georeferencing offset correction flag" ;
                georef_offset_corrected:modifiers = "" ;
        short image0(time, y, x) ;
                image0:_FillValue = -32767s ;
                image0:calibration = "reflectance" ;
                image0:description = "SEVIRI VIS006" ;
                image0:end_time = "2009-07-01 12:12:40.964000" ;
                image0:id_tag = "ch_r06" ;
                image0:long_name = "image0" ;
                image0:modifiers = "" ;
                image0:resolution = 3000.403165817 ;
                image0:standard_name = "toa_bidirectional_reflectance" ;
                image0:start_time = "2009-07-01 12:00:10.021000" ;
                image0:sun_earth_distance_correction_applied = "false" ;
                image0:sun_earth_distance_correction_factor = 1. ;
                image0:sun_zenith_angle_correction_applied = "false" ;
                image0:units = "%" ;
                image0:wavelength = 0.56, 0.635, 0.71 ;
                image0:coordinates = "acq_time lat lon" ;
                image0:add_offset = 0. ;
                image0:scale_factor = 0.01 ;
        short image1(time, y, x) ;
                image1:_FillValue = -32767s ;
                image1:calibration = "reflectance" ;
                image1:description = "SEVIRI VIS008" ;
                image1:end_time = "2009-07-01 12:12:40.964000" ;
                image1:id_tag = "ch_r09" ;
                image1:long_name = "image1" ;
                image1:modifiers = "" ;
                image1:resolution = 3000.403165817 ;
                image1:standard_name = "toa_bidirectional_reflectance" ;
                image1:start_time = "2009-07-01 12:00:10.021000" ;
                image1:sun_earth_distance_correction_applied = "false" ;
                image1:sun_earth_distance_correction_factor = 1. ;
                image1:sun_zenith_angle_correction_applied = "false" ;
                image1:units = "%" ;
                image1:wavelength = 0.74, 0.81, 0.88 ;
                image1:coordinates = "acq_time lat lon" ;
                image1:add_offset = 0. ;
                image1:scale_factor = 0.01 ;
        short image10(time, y, x) ;
                image10:_FillValue = -32767s ;
                image10:calibration = "brightness_temperature" ;
                image10:description = "SEVIRI WV_073" ;
                image10:end_time = "2009-07-01 12:12:40.964000" ;
                image10:id_tag = "ch_tb73" ;
                image10:long_name = "image10" ;
                image10:modifiers = "" ;
                image10:resolution = 3000.403165817 ;
                image10:standard_name = "brightness_temperature" ;
                image10:start_time = "2009-07-01 12:00:10.021000" ;
                image10:sun_earth_distance_correction_applied = "false" ;
                image10:sun_earth_distance_correction_factor = 1. ;
                image10:sun_zenith_angle_correction_applied = "false" ;
                image10:units = "K" ;
                image10:wavelength = 6.85, 7.35, 7.85 ;
                image10:coordinates = "acq_time lat lon" ;
                image10:add_offset = 273.15 ;
                image10:scale_factor = 0.01 ;
        short image2(time, y, x) ;
                image2:_FillValue = -32767s ;
                image2:calibration = "reflectance" ;
                image2:description = "SEVIRI IR_016" ;
                image2:end_time = "2009-07-01 12:12:40.964000" ;
                image2:id_tag = "ch_r16" ;
                image2:long_name = "image2" ;
                image2:modifiers = "" ;
                image2:resolution = 3000.403165817 ;
                image2:standard_name = "reflectance" ;
                image2:start_time = "2009-07-01 12:00:10.021000" ;
                image2:sun_earth_distance_correction_applied = "false" ;
                image2:sun_earth_distance_correction_factor = 1. ;
                image2:sun_zenith_angle_correction_applied = "false" ;
                image2:units = "%" ;
                image2:wavelength = 1.5, 1.64, 1.78 ;
                image2:coordinates = "acq_time lat lon" ;
                image2:add_offset = 0. ;
                image2:scale_factor = 0.01 ;
        short image3(time, y, x) ;
                image3:_FillValue = -32767s ;
                image3:calibration = "brightness_temperature" ;
                image3:description = "SEVIRI IR_039" ;
                image3:end_time = "2009-07-01 12:12:40.964000" ;
                image3:id_tag = "ch_tb37" ;
                image3:long_name = "image3" ;
                image3:modifiers = "" ;
                image3:resolution = 3000.403165817 ;
                image3:standard_name = "brightness_temperature" ;
                image3:start_time = "2009-07-01 12:00:10.021000" ;
                image3:sun_earth_distance_correction_applied = "false" ;
                image3:sun_earth_distance_correction_factor = 1. ;
                image3:sun_zenith_angle_correction_applied = "false" ;
                image3:units = "K" ;
                image3:wavelength = 3.48, 3.92, 4.36 ;
                image3:coordinates = "acq_time lat lon" ;
                image3:add_offset = 273.15 ;
                image3:scale_factor = 0.01 ;
        short image4(time, y, x) ;
                image4:_FillValue = -32767s ;
                image4:calibration = "brightness_temperature" ;
                image4:description = "SEVIRI IR_087" ;
                image4:end_time = "2009-07-01 12:12:40.964000" ;
                image4:id_tag = "ch_tb85" ;
                image4:long_name = "image4" ;
                image4:modifiers = "" ;
                image4:resolution = 3000.403165817 ;
                image4:standard_name = "brightness_temperature" ;
                image4:start_time = "2009-07-01 12:00:10.021000" ;
                image4:sun_earth_distance_correction_applied = "false" ;
                image4:sun_earth_distance_correction_factor = 1. ;
                image4:sun_zenith_angle_correction_applied = "false" ;
                image4:units = "K" ;
                image4:wavelength = 8.3, 8.7, 9.1 ;
                image4:coordinates = "acq_time lat lon" ;
                image4:add_offset = 273.15 ;
                image4:scale_factor = 0.01 ;
        short image5(time, y, x) ;
                image5:_FillValue = -32767s ;
                image5:calibration = "brightness_temperature" ;
                image5:description = "SEVIRI IR_108" ;
                image5:end_time = "2009-07-01 12:12:40.964000" ;
                image5:id_tag = "ch_tb11" ;
                image5:long_name = "image5" ;
                image5:modifiers = "" ;
                image5:resolution = 3000.403165817 ;
                image5:standard_name = "brightness_temperature" ;
                image5:start_time = "2009-07-01 12:00:10.021000" ;
                image5:sun_earth_distance_correction_applied = "false" ;
                image5:sun_earth_distance_correction_factor = 1. ;
                image5:sun_zenith_angle_correction_applied = "false" ;
                image5:units = "K" ;
                image5:wavelength = 9.8, 10.8, 11.8 ;
                image5:coordinates = "acq_time lat lon" ;
                image5:add_offset = 273.15 ;
                image5:scale_factor = 0.01 ;
        short image6(time, y, x) ;
                image6:_FillValue = -32767s ;
                image6:calibration = "brightness_temperature" ;
                image6:description = "SEVIRI IR_120" ;
                image6:end_time = "2009-07-01 12:12:40.964000" ;
                image6:id_tag = "ch_tb12" ;
                image6:long_name = "image6" ;
                image6:modifiers = "" ;
                image6:resolution = 3000.403165817 ;
                image6:standard_name = "brightness_temperature" ;
                image6:start_time = "2009-07-01 12:00:10.021000" ;
                image6:sun_earth_distance_correction_applied = "false" ;
                image6:sun_earth_distance_correction_factor = 1. ;
                image6:sun_zenith_angle_correction_applied = "false" ;
                image6:units = "K" ;
                image6:wavelength = 11., 12., 13. ;
                image6:coordinates = "acq_time lat lon" ;
                image6:add_offset = 273.15 ;
                image6:scale_factor = 0.01 ;
        short image7(time, y, x) ;
                image7:_FillValue = -32767s ;
                image7:calibration = "brightness_temperature" ;
                image7:description = "SEVIRI IR_134" ;
                image7:end_time = "2009-07-01 12:12:40.964000" ;
                image7:id_tag = "ch_tb133" ;
                image7:long_name = "image7" ;
                image7:modifiers = "" ;
                image7:resolution = 3000.403165817 ;
                image7:standard_name = "brightness_temperature" ;
                image7:start_time = "2009-07-01 12:00:10.021000" ;
                image7:sun_earth_distance_correction_applied = "false" ;
                image7:sun_earth_distance_correction_factor = 1. ;
                image7:sun_zenith_angle_correction_applied = "false" ;
                image7:units = "K" ;
                image7:wavelength = 12.4, 13.4, 14.4 ;
                image7:coordinates = "acq_time lat lon" ;
                image7:add_offset = 273.15 ;
                image7:scale_factor = 0.01 ;
        short image8(time, y, x) ;
                image8:_FillValue = -32767s ;
                image8:calibration = "brightness_temperature" ;
                image8:description = "SEVIRI IR_097" ;
                image8:end_time = "2009-07-01 12:12:40.964000" ;
                image8:id_tag = "ch_tb97" ;
                image8:long_name = "image8" ;
                image8:modifiers = "" ;
                image8:resolution = 3000.403165817 ;
                image8:standard_name = "brightness_temperature" ;
                image8:start_time = "2009-07-01 12:00:10.021000" ;
                image8:sun_earth_distance_correction_applied = "false" ;
                image8:sun_earth_distance_correction_factor = 1. ;
                image8:sun_zenith_angle_correction_applied = "false" ;
                image8:units = "K" ;
                image8:wavelength = 9.38, 9.66, 9.94 ;
                image8:coordinates = "acq_time lat lon" ;
                image8:add_offset = 273.15 ;
                image8:scale_factor = 0.01 ;
        short image9(time, y, x) ;
                image9:_FillValue = -32767s ;
                image9:calibration = "brightness_temperature" ;
                image9:description = "SEVIRI WV_062" ;
                image9:end_time = "2009-07-01 12:12:40.964000" ;
                image9:id_tag = "ch_tb67" ;
                image9:long_name = "image9" ;
                image9:modifiers = "" ;
                image9:resolution = 3000.403165817 ;
                image9:standard_name = "brightness_temperature" ;
                image9:start_time = "2009-07-01 12:00:10.021000" ;
                image9:sun_earth_distance_correction_applied = "false" ;
                image9:sun_earth_distance_correction_factor = 1. ;
                image9:sun_zenith_angle_correction_applied = "false" ;
                image9:units = "K" ;
                image9:wavelength = 5.35, 6.25, 7.15 ;
                image9:coordinates = "acq_time lat lon" ;
                image9:add_offset = 273.15 ;
                image9:scale_factor = 0.01 ;
        double projection_area_extent(time, corners) ;
                projection_area_extent:_FillValue = NaN ;
                projection_area_extent:comment = "Area extent (lower left x, lower left y, upper right x, upper right y) in projection coordinates" ;
                projection_area_extent:long_name = "Projection area extent" ;
                projection_area_extent:modifiers = "" ;
                projection_area_extent:units = "meters" ;
        double satellite_altitude(time) ;
                satellite_altitude:_FillValue = NaN ;
                satellite_altitude:comment = "Relative and perpendicular to the surface of the earth ellipsoid. Not to be confused with projection altitude." ;
                satellite_altitude:long_name = "Actual satellite altitude" ;
                satellite_altitude:modifiers = "" ;
                satellite_altitude:units = "meters" ;
        double satellite_latitude(time) ;
                satellite_latitude:_FillValue = NaN ;
                satellite_latitude:comment = "Not to be confused with projection latitude" ;
                satellite_latitude:long_name = "Actual satellite latitude" ;
                satellite_latitude:modifiers = "" ;
                satellite_latitude:units = "degree_north" ;
        double satellite_longitude(time) ;
                satellite_longitude:_FillValue = NaN ;
                satellite_longitude:comment = "Not to be confused with projection longitude" ;
                satellite_longitude:long_name = "Actual satellite longitude" ;
                satellite_longitude:modifiers = "" ;
                satellite_longitude:units = "degree_east" ;
        short satzenith(time, y, x) ;
                satzenith:_FillValue = -32767s ;
                satzenith:end_time = "2009-07-01 12:12:40.964000" ;
                satzenith:id_tag = "satzenith" ;
                satzenith:long_name = "satellite zenith angle" ;
                satzenith:modifiers = "" ;
                satzenith:standard_name = "sensor_zenith_angle" ;
                satzenith:start_time = "2009-07-01 12:00:10.021000" ;
                satzenith:units = "degree" ;
                satzenith:valid_range = 0LL, 9000LL ;
                satzenith:coordinates = "acq_time lat lon" ;
                satzenith:add_offset = 0. ;
                satzenith:scale_factor = 0.01 ;
        short sunzenith(time, y, x) ;
                sunzenith:_FillValue = -32767s ;
                sunzenith:end_time = "2009-07-01 12:12:40.964000" ;
                sunzenith:id_tag = "sunzenith" ;
                sunzenith:long_name = "sun zenith angle" ;
                sunzenith:modifiers = "" ;
                sunzenith:standard_name = "solar_zenith_angle" ;
                sunzenith:start_time = "2009-07-01 12:00:10.021000" ;
                sunzenith:units = "degree" ;
                sunzenith:valid_range = 0LL, 18000LL ;
                sunzenith:coordinates = "acq_time lat lon" ;
                sunzenith:add_offset = 0. ;
                sunzenith:scale_factor = 0.01 ;
        double time_bnds(time, bnds_1d) ;
                time_bnds:units = "days since 2004-01-01" ;
                time_bnds:calendar = "standard" ;

// global attributes:
                :date_created = "2019-11-25T17:50:17Z" ;
                :end_time = "2009-07-01 12:15:00" ;
                :image_rotated = "true" ;
                :instrument = "SEVIRI" ;
                :orbit_number = "99999" ;
                :platform = "Meteosat-9" ;
                :projection = "geos" ;
                :projection_altitude = 35785831. ;
                :projection_latitude = 0. ;
                :projection_longitude = 0.f ;
                :projection_semi_major_axis = 6378169. ;
                :projection_semi_minor_axis = 6356583.8 ;
                :source = "seviri2pps.py" ;
                :start_time = "2009-07-01 12:00:00" ;
                :history = "Created by pytroll/satpy on 2019-11-25 17:50:17.076401" ;
                :Conventions = "CF-1.7" ;
```

</details>

This should contain everything we need for CLAAS level 2 products. @ninahakansson Can we tell PPS to  transfer selected global attributes and variables to its output? Otherwise no problem, we can add a post-processing step for that.

- [X] Improve time encoding in seviri2pps
- [x] Store non-static metadata in nc variables

Depends on https://github.com/pytroll/satpy/pull/976

sfinkens committed:
Improve time encoding in seviri2pps

sfinkens committed:
Make time dimension unlimited

sfinkens committed:
Store non-static metadata in nc variables

sfinkens committed:
Add tests

sfinkens committed:
Remove unused import

ninahakansson commented:
Thanks for the PR @sfinkens!

In the ncdump it seem _FillValue and valid_range have different datatypes, should they not be the same? But this is not new in the PR I suspect.

Do I get it right that you would want some variables and attributes from level1c appended to the cloud probability and ctth files? I would be possible, but PPS reading (and writing) is much focused on reading (writing) only what is needed.

I will try to run PPS on a file with the unlimited time dimension.

sfinkens committed:
Fix units of angles' valid range attribute

sfinkens committed:
Update tests

sfinkens commented:
Yes, that was my idea. But I see your point and I wouldn't want to trigger PPS changes right now. We will be doing a post-processing anyway to make PPS & CPP products compliant with the CM SAF metadata conventions. Then we can also collect some information from the level 1c files - if they contain all the information we need. Of course the extra information I added in this PR shouldn't cause problems in PPS. I will ask Imke to test it. Here is an example file if you want to have a look as well: https://public.cmsaf.dwd.de/data/sfinkens/level1c4pps/

P.S. I fixed the valid range units.

sfinkens commented on level1c4pps/gac2pps_lib.py:
```diff
@@ -87,7 +87,7 @@ def update_ancilliary_datasets(scene):
     scanline_timestamps = np.array(scene['qual_flags'].coords['acq_time'] -
                                    first_jan_1970).astype(dtype='timedelta64[ms]').astype(np.float64)
-    scene['scanline_timestamps'] = xr.DataArray(da.from_array(scanline_timestamps),
+    scene['scanline_timestamps'] = xr.DataArray(da.from_array(scanline_timestamps, chunks=1024),
```
In more recent dask versions you have to specify a chunksize. I chose 1024, then you have roughly 12 chunks per file.

ninahakansson replied:
Good to know that chunksizes will have to be specified!

ninahakansson commented on level1c4pps/__init__.py:
```diff
@@ -134,22 +139,24 @@ def get_band_encoding(dataset, bandnames, pps_tagnames, chunks=None):
         if chunks is not None:
             enc['chunksizes'] = chunks
-    if name in ['lon', 'lat']:
+    elif name in ['lon', 'lat']:
```
I would like to have an if here and not elif. To make it possible to add more id_tags for example to lat/lon without changing this function. The qual_flags  (gac) I think also have the id_tag attribute, and with this update they would not get the correct encoding. However my test was to badly written to capture it.

I should have defined qual_f as:
qual_f = mock.MagicMock(attrs={'name': 'qual_flags',
                                                     'id_tag': 'qual_flags'})

sfinkens replied:
Oh, sorry! I have reverted the changes as you suggested.

sfinkens committed:
Revert erroneous changes to band encoding

ninahakansson reviewed (approved):
Looks good, thanks for fixing the broken test!
